File: firebase/functions/lib/community/unsubscribe_from_community_notifications.dart
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/infra/firestore_utils.dart';
import '../utils/notifications_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/community/community_user_settings.dart';
import 'package:data_models/community/membership.dart';

class UnsubscribeFromCommunityNotifications
    extends OnCallMethod<UnsubscribeFromCommunityNotificationsRequest> {
  NotificationsUtils notificationsUtils;

  UnsubscribeFromCommunityNotifications({
    NotificationsUtils? notificationsUtils,
  })  : notificationsUtils = notificationsUtils ?? NotificationsUtils(),
        super(
          'unsubscribeFromCommunityNotifications',
          (jsonMap) =>
              UnsubscribeFromCommunityNotificationsRequest.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    UnsubscribeFromCommunityNotificationsRequest request,
    CallableContext context,
  ) async {
    final decrypted = notificationsUtils.decryptUnsubscribeData(request.data);

    final membershipsData = await firestore
        .collectionGroup('community-membership')
        .where(Membership.kFieldUserId, isEqualTo: decrypted.userId)
        .whereNotEqual(Membership.kFieldStatus, notEqualTo: 'nonmember')
        .get();

    final communityIds = membershipsData.documents
        .map(
          (e) => Membership.fromJson(
            firestoreUtils.fromFirestoreJson(e.data.toMap()),
          ).communityId,
        )
        .toList();

    await Future.wait(
      communityIds.map((communityId) {
        final settings = CommunityUserSettings(
          userId: decrypted.userId,
          communityId: communityId,
          notifyAnnouncements: NotificationEmailType.none,
          notifyEvents: NotificationEmailType.none,
        );

        return firestore
            .document(
              'privateUserData/${decrypted.userId}/communityUserSettings/$communityId',
            )
            .setData(
              DocumentData.fromMap(
                firestoreUtils.toFirestoreJson(settings.toJson()),
              ),
              SetOptions(merge: true),
            );
      }),
    );
  }
}

File: firebase/functions/src/community/unsubscribe_from_community_notifications.ts
import * as functions from 'firebase-functions'
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils'
import { OnCallMethod } from '../on_call_function'
import {
    UnsubscribeFromCommunityNotificationsRequest,
    Membership,
    MembershipStatus,
    CommunityUserSettings,
    NotificationEmailType,
} from '../types'
import { decryptUnsubscribeData } from '../utils/encrypt_util'

export class UnsubscribeFromCommunityNotifications extends OnCallMethod<UnsubscribeFromCommunityNotificationsRequest> {
    constructor() {
        super(
            'unsubscribeFromCommunityNotifications',
            (json) => json as UnsubscribeFromCommunityNotificationsRequest
        )
    }

    async action(
        request: UnsubscribeFromCommunityNotificationsRequest,
        _context: functions.https.CallableContext
    ): Promise<void> {
        const decrypted = decryptUnsubscribeData(request.data)

        const membershipsData = await firestore
            .collectionGroup('community-membership')
            .where('userId', '==', decrypted.userId)
            .where('status', '!=', MembershipStatus.nonmember)
            .get()

        const communityIds = membershipsData.docs.map(
            (doc) => (doc.data() as Membership).communityId
        )

        await Promise.all(
            communityIds.map((communityId) => {
                const settings: CommunityUserSettings = {
                    userId: decrypted.userId,
                    communityId,
                    notifyAnnouncements: NotificationEmailType.none,
                    notifyEvents: NotificationEmailType.none,
                }

                return firestore
                    .doc(`privateUserData/${decrypted.userId}/communityUserSettings/${communityId}`)
                    .set(
                        firestoreUtils.toFirestoreJson(
                            settings as unknown as Record<string, unknown>
                        ),
                        { merge: true }
                    )
            })
        )
    }
}
